src/Microsoft.ML.Tokenizers/Model/BPE.cs

```diff
@@ -0,0 +1,448 @@
+            ContinuingSubwordPrefix = continuingSubwordPrefix;
+            EndOfWordSuffix = endOfWordSuffix;
+
+            (Dictionary<string, int>? vocab1, Vec<(string, string)> merges) = ReadFile(vocabFile, mergesFile);
+            Vocab = vocab1 ?? new Dictionary<string, int>();
+
+            VocabReverse = new();
+
+            foreach (KeyValuePair<string, int> kvp in Vocab)
+            {
+                VocabReverse.Add(kvp.Value, kvp.Key);
+            }
+
+            if (unknownToken is null && VocabReverse.TryGetValue(0, out string unkToken))
+            {
+                unknownToken = unkToken;
+            }
+
+            UnknownToken = unknownToken;
+
+            int prefixLen = ContinuingSubwordPrefix is null ? 0 : ContinuingSubwordPrefix.Length;
+
+            Merges = new();
+            for (int i = 0; i < merges.Count; i++)
+            {
+                (string a, string b) mergeValues = merges[i];
+
+                if (!Vocab.TryGetValue(mergeValues.a, out int aId))
+                {
+                    throw new InvalidOperationException($"Trying to merge a token {mergeValues.a} which not exist in the vocabulary.");
+                }
+
+                if (!Vocab.TryGetValue(mergeValues.b, out int bId))
+                {
+                    throw new InvalidOperationException($"Trying to merge a token {mergeValues.b} which not exist in the vocabulary.");
+                }
+
+                string newToken = $"{mergeValues.a}{mergeValues.b.Substring(prefixLen)}";
+                if (!Vocab.TryGetValue(newToken, out int newId))
+                {
+                    throw new InvalidOperationException($"Trying to merge a token {newToken} which not exist in the vocabulary.");
+                }
+
+                Merges.Add(new Pair<int>(aId, bId), (i, newId));
+            }
+        }
+
+        /// <summary>
+        /// Gets the Bpe decoder object.
+        /// </summary>
+        public static TokenizerDecoder Decoder { get; } = new BpeDecoder();
+
+        /// <summary>
+        /// Tokenize a sequence string to a list of tokens.
+        /// </summary>
+        /// <param name="sequence">The sequence to tokenize.</param>
+        /// <returns>The list of tokens generated from the sequence tokenization.</returns>
+        public override IReadOnlyList<Token> Tokenize(string sequence)
+        {
+            if (sequence.Length == 0)
+            {
+                return EmptyTokensList;
+            }
+
+            if (!Dropout.HasValue)
+            {
+                return TokenizeWithCache(sequence);
+            }
+
+            Word word = MergeWord(sequence);
+
+            return WordToTokens(ref word);
+        }
+
+        /// <summary>
+        /// Map the token to tokenized Id.
+        /// </summary>
+        /// <param name="token">The token to map to the Id.</param>
+        /// <returns>The mapped Id of the token.</returns>
+        public override int? TokenToId(string token)
```

Comment:
Is there a reason this takes a string when we have an actual `Token` object?

Reply:
Nevermind, I see the implicit conversion and that `Vocab` is from string to int.